vcd/resource_vcd_vapp_nat_rules.go

```diff
@@ -0,0 +1,306 @@
+		vappNetwork.Configuration.Features.NatService.Policy == allowTrafficPolicy {
+		_ = d.Set("enable_ip_masquerade", false)
+	}
+	_ = d.Set("nat_type", vappNetwork.Configuration.Features.NatService.NatType)
+	_ = d.Set("rule", rules)
+	return nil
+}
+
+func expandVappNetworkNatRules(d *schema.ResourceData, vapp *govcd.VApp, natType string) ([]*types.NatRule, error) {
+
+	var natRules []*types.NatRule
+	for _, singleRule := range d.Get("rule").([]interface{}) {
+		configuredRule := singleRule.(map[string]interface{})
+		if natType == portForwardingNatType {
+			rule := &types.NatRule{
+				VMRule: &types.NatVMRule{
+					ExternalPort: configuredRule["external_port"].(int),
+					VMNicID:      configuredRule["vm_nic_id"].(int),
+					InternalPort: configuredRule["forward_to_port"].(int),
+					Protocol:     configuredRule["protocol"].(string),
+				},
+			}
+			vm, err := vapp.GetVMById(configuredRule["vm_id"].(string), false)
+			if err != nil {
+				return nil, fmt.Errorf("error fetchining VM: %s", err)
+			}
+			rule.VMRule.VAppScopedVMID = vm.VM.VAppScopedLocalID
+			natRules = append(natRules, rule)
+		} else if natType == ipTranslationNatType {
+			rule := &types.NatRule{
+				OneToOneVMRule: &types.NatOneToOneVMRule{
+					MappingMode: configuredRule["mapping_mode"].(string),
+					VMNicID:     configuredRule["vm_nic_id"].(int),
+				},
+			}
+			externalIp := configuredRule["external_ip"].(string)
+			if externalIp != "" {
+				rule.OneToOneVMRule.ExternalIPAddress = &externalIp
+			}
+
+			vm, err := vapp.GetVMById(configuredRule["vm_id"].(string), false)
+			if err != nil {
+				return nil, fmt.Errorf("error fetchining VM: %s", err)
+			}
+			rule.OneToOneVMRule.VAppScopedVMID = vm.VM.VAppScopedLocalID
+			natRules = append(natRules, rule)
+		}
+	}
+
+	return natRules, nil
+}
+
+// vappNetworkNatRuleImport is responsible for importing the resource.
+// The following steps happen as part of import
+// 1. The user supplies `terraform import _resource_name_ _the_id_string_` command
+// 2. `_the_id_string_` contains a dot formatted path to resource as in the example below
+// 3. The functions splits the dot-formatted path and tries to lookup the object
+// 4. If the lookup succeeds it set's the ID field for `_resource_name_` resource in state file
+// (the resource must be already defined in .tf config otherwise `terraform import` will complain)
+// 5. `terraform refresh` is being implicitly launched. The Read method looks up all other fields
+// based on the known ID of object.
+//
+// Example resource name (_resource_name_): vcd_vapp_nat_rules.my_existing_nat_rules
+// Example import path (_the_id_string_): org.my_existing_vdc.vapp_name.network_name or org.my_existing_vdc.vapp_id.network_id
+// Note: the separator can be changed using Provider.import_separator or variable VCD_IMPORT_SEPARATOR
+func vappNetworkNatRuleImport(d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
+	return vappFirewallRuleImport(d, meta)
```

Comment:
At least name and debug messages look to be invalid and confusing when `vappFirewallRuleImport` are used. If this is duplicate functionality I suggest to improve naming and logging.

Reply:
+1. If re-use is wanted, then there should be a generic function which both `vappNetworkNatRuleImport` and `vappFirewallRuleImport` call upon. Now it looks as if NAT rules are based on FW rules.

Reply:
done